crates/glue/writing_a_glue_plugin.md

```diff
@@ -0,0 +1,441 @@
+
+## Struct
+
+A Roc record type. Also known as a struct or a product type. Its `fields` field can be one of two variants:
+
+### HasNoClosure
+
+Contains a list of fields with a name and type. Fields are arranged in memory in listed order as they would be in a C struct. Fields are not guaranteed to be in order of declaration in Roc.
+
+```roc
+# in `platform.roc`
+Color : { red : U8, green : U8, blue : U8 }
+```
+```roc
+# shape reported by `roc glue`
+Struct {
+    name: "Color",
+    fields: (HasNoClosure [
+        # `@TypeId 1` == `Num U8`
+        {id: (@TypeId 1), name: "blue"},
+        {id: (@TypeId 1), name: "green"},
+        {id: (@TypeId 1), name: "red"},
+    ]),
+}
+```
+```c
+// example C layout
+struct Color {
+    uint8_t blue;
+    uint8_t green;
+    uint8_t red;
+};
+```
+
+### HasClosure
+
+(TODO: no glue spec currently supports this. Unsure if this either does not work or if nobody who knows how has gotten around to it.)
+
+## TagUnion
+
+Roc's tagged union/sum types. Roc may choose to optimize a tag union's in-memory representation behind the scenes. This results in tag unions coming in seven different flavors.
+
+Some info common across all of these are:
+- All of these are guaranteed have at least one tag (see also: [EmptyTagUnion](#emptytagunion)).
+- Discriminants (when stored in a discriminant field or as tags in a tagged pointer) are equal to a tag's index in the list the variants are given in (e.g: the `tags` field in an `Enumeration`) (if applicable).
+- Discriminants (when stored in a discriminant field) are currently either an 8-bit or 16-bit unsigned integer.
+
+### NonRecursive
+
+The base case: a tag union with more than one tag which does not contain itself, even through indirection.
+
+```roc
+# in `platform.roc`
+Cell : [
+    Empty,
+    Number I32,
+    Text Str,
+]
+```
+```roc
+# shape reported by `roc glue`
+TagUnion (NonRecursive {
+    name: "Cell",
+    discriminantSize: 1,
+    discriminantOffset: 24,
+    tags: [
+        {name: "Empty", payload: None},
+        {name: "Number", payload: (Some (@TypeId 1))},
+        {name: "Text", payload: (Some (@TypeId 2))}
+    ]
+})
+```
+```c
+// example C layout
+// note: it is currently guaranteed that the discriminant comes after the
+// space reserved in memory for the union, including padding. this means
+// NonRecursive tag unions may be represented as a struct, as below. this may
+// change in the future, as this currently makes for a missed optimization:
+// https://github.com/roc-lang/roc/issues/7171
+struct Cell {
+    union {
+        int32_t Number;
+        RocStr Text;
+    } payload;
+    enum {
+        Empty, Number, Text
+    } discriminant;
+};
+```
+
+### Enumeration
+
+A special case where a none of the tags in a tag union contains a payload. This means it can be represented as just its discriminant.
```

Comment:
```suggestion
A special case where none of the tags in a tag union contains a payload. This means it can be represented as just its discriminant.
```